### DISCUSSION TITLE
How to overlay features in the order determined by its properties in GeoJsonLayer?

### DISCUSSION THREAD

**Discussion by fifzteen** [visgl/deck.gl#6152] [Q&A]
Hi.
In GeoJsonLayer, the subLayers overlaid in the order `polygons-fill`, `polygons-stroke`, `linstrings`, `point-xxx`.

In loaded MVT, geojson data has layerName.

```
{ type: 'Feature', properties: { layerName: 'foo' }, geometry: {...} }
```

I have the orders of these layers in MVT,
and I want to overlay each polygon features in the order determined by its own layerName.
How can I overlay features like below?

![overlay_by_layerName_order](https://user-images.githubusercontent.com/25410805/131602117-4f027eb7-30b3-4763-bf43-12350c2e0837.png)

---

I used to create GeoJsonLayers for each layer in mvt at `renderSubLayers`.

```
new MVTLayer({
  ...,
  parameters: { depthTest: false },
  renderSubLayers: (props) => [
    new GeoJsonLayer({...}), // For Layer1
    new GeoJsonLayer({...}), // For Layer2
    ...
  ],
})
```

This way is good about overlaying layers, but creates too many pickable layers.
For example,
If the view covers 9 tiles and 10 pickable GeoJsonLayers in each mvt tile, The count of all pickable layer is 9 * 10 * 4 = 360. over 255.
so I'm tring to draw all layers in mvt in one GeoJsonLayer.

---

I try adding small z to each coordinates.
https://codepen.io/fifzteen/pen/OJgMQYm

The paths behind upper 'polygons-fill' feature are not drawn (only not shown?)
Setting `blendFunc` solves this problem?

**Comment by Pessimistress** [comment]
The best way to handle this is to split the features into multiple sublayers and layers are always drawn in order, as you have been doing. Once #6149 is resolved it should improve your use case, though there still are rooms for improvement:

### Index pickable layers more efficiently 

Looks like the warning is issued if the number of primitive layers with `props.visible && props.pickable` exceeds 255:

https://github.com/visgl/deck.gl/blob/v8.5.5/modules/core/src/lib/deck-picker.js#L134

This does not account for dynamic, viewport-specific visibility calculated from `layerFilter` or `parent.filterSubLayer`, which is used by the TileLayer/MVTLayer. Here `layerIndex` is incremented even if `shouldDrawLayer` is `false`:

https://github.com/visgl/deck.gl/blob/v8.5.5/modules/core/src/passes/layers-pass.js#L89

And subsequently used for the alpha channel of the picking color:

https://github.com/visgl/deck.gl/blob/v8.5.5/modules/core/src/passes/pick-layers-pass.js#L93

This means that any tile in the tile cache, even if outside of the current viewport, still takes up slots in the limited picking stack. The default tile cache is [5x](https://deck.gl/docs/api-reference/geo-layers/tile-layer#maxcachesize) what's visible so you only get 20% of the 255 limit once the cache is filled up. Since `filterSubLayer` is new in v8.5 I believe this behavior is a regression and we should be able to fix it easily.

### Use additional channels to encode layer index

We have a `layerRenderIndex` that resolves to the same value for the same sublayer in each tile (see #4388) so we could theoretically encode the picking colors across multiple tiles with the same layer index. That would alter the limit to 65535 objects per layer per tile x 255 tiles x 255 layers per tile which seems a lot more reasonable.

The implementation of this will involve calculating the secondary index and adding some new uniforms to "offset" the picking color in the shader. I think it's definitely doable, but will be hard to test and prone to breakage due to the additional complexity.

cc @alasarr and @ibgreen if this is of interest to your teams as well I can put together an RFC for 8.6.

**Reply by fifzteen** [reply]
Thank you for your reply.

> ### Index pickable layers more efficiently

Without fixing, I can apply tile visible state to pickable states of its sublayers, like "Implementetion A" in `CompositeLayer.filterSubLayers`, isn't it? (I will try this)

```
new MVTLayer({
  renderSubLayers: (props) => [
    new GeoJsonLayer({
      pickable: props.tile.visible && (the pickable value originally set)
    }), // Layer1
    ... // Layer2 and more
  ]
})
```

To fix, Is it reasonable to add checking `filterSubLayers` consideration in `DeckPicker._getPickable()`?

```
const pickableLayers = layers.filter(layer => 
  layer.isPickable() && 
  !layer.isComposite && 
  (!layer.parent.isComposite || 
    viewports.some((viewport) => 
      layer.parent.filterSubLayers({
        layer,
        viewport,
        isPicking: true,
        pass: this.pickLayersPass.id
      })
    )
  )
);
```

> ### Use additional channels to encode layer index

The Mapbox Vector Tile Specification states that a tile should have at least one layer.(That means a tile can have multiple layers)
so, other deck.gl users will have similar problems.
I think it would be GREAT that MVTLayer can easily handle layers in MVT (overlaying, picking, ...and so on) without hacks.

**Reply by Pessimistress** [reply]
> To fix, Is it reasonable to add checking filterSubLayers consideration in DeckPicker._getPickable()?

It's more complicated than that because `filterSubLayers` is recursive, and `layerFilter` can affect visibility as well. I think we should avoid duplicating the `shouldDrawLayer` logic (which undergoes a breaking change in 8.6) in multiple places and always defer to the render pass to determine which layers are pickable. I will open a PR to properly fix this.

Considering this is touching more code than I originally thought, I'm leaning towards only including the fix in v8.6. We are currently looking at a release date at the end of September.

**Reply by fifzteen** [reply]
> Considering this is touching more code than I originally thought, I'm leaning towards only including the fix in v8.6. We are currently looking at a release date at the end of September.

I got it. Thank you for getting on with the fix.
I'm looking forward to the v8.6 release!
